nbextensions/bulkImportCell/bulkImportCell.js

```diff
@@ -0,0 +1,186 @@
+define([
+    'uuid',
+    'common/appUtils',
+    'common/utils',
+    'common/runtime',
+    'common/busEventManager',
+    'base/js/namespace'
+], (
+    Uuid,
+    AppUtils,
+    Utils,
+    Runtime,
+    BusEventManager,
+    Jupyter
+) => {
+    'use strict';
+    const CELL_TYPE = 'app-bulk-import';
+
+    /**
+     * This class creates and manages the bulk import cell. This works with, and wraps around,
+     * the Jupyter Cell object.
+     */
+    class BulkImportCell {
+        /**
+         * The constructor does the work of initializing the bulk import cell module. It
+         * modifies the cell metadata, if the initialize parameter is truthy, and extends the
+         * cell object to do things we need it to, like respond to minimization requests and
+         * render its icon. This modification also includes the contents of typesToFiles if
+         * present.
+         *
+         * If initialize is falsy, no changes are made to the structure, even if typesToFiles is
+         * present and contains new data.
+         * @param {Cell} cell a Jupyter notebook cell. This should be a code cell, and will throw
+         * an Error if it is not.
+         * @param {boolean} initialize if true, this will initialize the bulk import cell
+         * structure that gets serialized in the cell metadata. This should ONLY be set true when
+         * creating a new bulk import cell, not loading a narrative that already contains one
+         * @param {object} typesToFiles keys = data types, values = arrays of files to import
+         */
+        constructor(cell, initialize, typesToFiles) {
+            if (cell.cell_type !== 'code') {
+                throw new Error('Can only create Bulk Import Cells out of code cells!');
+            }
+            this.cell = cell;
+            this.kbaseNode = null;  // this is the
+            this.runtime = Runtime.make();
+            this.cellBus = null;
+            this.busEventManager = BusEventManager.make({
+                bus: this.runtime.bus()
+            });
+            if (initialize) {
+                this.initialize(typesToFiles);
+            }
+            this.setupCell();
+        }
+
+        /**
+         * Returns true if the given cell should be treated as a bulk import cell
+         * @param {Cell} cell - a Jupyter Notebook cell
+         */
+        static isBulkImportCell(cell) {
+            if (cell.cell_type !== 'code' || !cell.metadata.kbase) {
+                return false;
+            }
+            return cell.metadata.kbase.type === CELL_TYPE;
+        }
+
+        /**
+         * Does the initial pass on newly created cells to initialize its metadata and get it
+         * set up for a new life as a Bulk Import Cell.
+         *
+         * @param {object} typesToFiles a map from object type to an array of files.
```

Comment:
doc note: can you specify what object type you mean here? kbase object type? data type?

Reply:
I updated that with an example.